toolbox/src/DiscreteFilter.cpp

```diff
@@ -0,0 +1,261 @@
+            if (error) {
+                error->message = ClassName
+                                 + " (FirstOrderLowPassFilter) You need to "
+                                   "specify Fc and Ts";
+            }
+            return false;
+        }
+        filter = new FirstOrderLowPassFilter(firstOrderLowPassFilter_fc.floatData(),
+                                             firstOrderLowPassFilter_ts.floatData());
+    }
+    // MedianFilter
+    // ------------
+    else if (filter_type == "MedianFilter") {
+        if (medianFilter_order.int32Data() == 0) {
+            if (error) {
+                error->message = ClassName
+                                 + " (MedianFilter) You need to specify the "
+                                   "filter order.";
+            }
+            return false;
+        }
+        filter = new MedianFilter(medianFilter_order.int32Data());
+    }
+    else {
+        if (error) error->message = ClassName + " Filter type not recognized.";
+        return false;
+    }
+
+    return true;
+}
+
+bool DiscreteFilter::terminate(BlockInformation* blockInfo, wbt::Error* error)
+{
+    // Deallocate all the memory
+    // -------------------------
+
+    if (filter) {
+        delete filter;
+        filter = nullptr;
+    }
+
+    if (num) {
+        delete num;
+        num = nullptr;
+    }
+
+    if (den) {
+        delete den;
+        den = nullptr;
+    }
+
+    if (inputSignalVector) {
+        delete inputSignalVector;
+        inputSignalVector = nullptr;
+    }
+
+    return true;
+}
+
+bool DiscreteFilter::output(BlockInformation* blockInfo, wbt::Error* error)
+{
+    if (filter == nullptr) return false;
+
+    // Get the input and output signals
+    Signal inputSignal  = blockInfo->getInputPortSignal(INPUT_IDX_SIGNAL);
+    Signal outputSignal = blockInfo->getOutputPortSignal(OUTPUT_IDX_SIGNAL);
+
+    unsigned inputSignalWidth = blockInfo->getInputPortWidth(INPUT_IDX_SIGNAL);
+
+    // Allocate the memory for the input data
+    if (firstRun) {
+        // Allocate the input signal
+        inputSignalVector = new Vector(inputSignalWidth, 0.0);
+
+        // Initialize the filter. This is required because if the signal is not 1D,
+        // the default filter constructor initialize a wrongly sized y0
+        filter->init(Vector(inputSignalWidth));
+
+        firstRun = false;
+    }
+
+    // Copy the Signal to the data structure that the filter wants
+    for (unsigned i = 0; i < inputSignalWidth; ++i) {
+        (*inputSignalVector)[i] = inputSignal.get(i).doubleData();
+    }
+
+    // Filter the current component of the input signal
+    const Vector& outputVector = filter->filt(*inputSignalVector);
+
+    // Forward the filtered signals to the output port
+    outputSignal.setBuffer(outputVector.data(), outputVector.length());
+
+    return true;
+}
+
+void DiscreteFilter::stringToYarpVector(const std::string str, Vector* v)
+{
+    assert(v != nullptr);
```

Comment:
As the pointer should always be `!=` null, I would use a reference, not a pointer